### PR TITLE
[TOPI][GL] OpenGL topi.

### PR DESCRIPTION
**Pull request by phisiart**
- [x] conv2d_nchw
- [x] dense
- [x] injective
- [x] pool
- [x] softmax

### CODE REVIEW

**Comment by phisiart** [comment]
Sorry I'll have to pause for ~ half a week. Being busy currently.

**Comment by phisiart** [comment]
I have encountered a problem that is essentially blocking OpenGL codegen.

If we write this program:

```python
m = 2
n = 2
p = 2
A = tvm.placeholder((m, n, p), name='A', dtype="int32")
B = tvm.compute((m, n, p), lambda *i: A[i], name="B")

s = tvm.create_schedule(B.op)
print(tvm.lower(s, [A, B], simple_mode=True))
s[B].opengl()
...
```

This program lowers to:

```
B[(((((threadIdx.x/4)*2) + ((threadIdx.x/2) % 2))*2) + (threadIdx.x % 2))] = A[(((((threadIdx.x/4)*2) + ((threadIdx.x/2) % 2))*2) + (threadIdx.x % 2))]
```

However, it actually should be simplified to:

```
B[threadIdx.x] = A[threadIdx.x]
```

It turns out that the following 2 programs can be properly simplified:

```python
# 2D tensor
m = 2
n = 2
A = tvm.placeholder((m, n), name='A', dtype="int32")
B = tvm.compute((m, n), lambda *i: A[i], name="B")

s = tvm.create_schedule(B.op)
print(tvm.lower(s, [A, B], simple_mode=True))
s[B].opengl()
...
```

```python
# 3D tensor but with variable shape
m = tvm.var("m")
n = tvm.var("n")
p = tvm.var("p")
A = tvm.placeholder((m, n, p), name='A', dtype="int32")
B = tvm.compute((m, n, p), lambda *i: A[i], name="B")

s = tvm.create_schedule(B.op)
print(tvm.lower(s, [A, B], simple_mode=True))
s[B].opengl()
...
```

I figured out that the reason is that `StorageFlatten` is not smart enough to simplify the index expression.

I think currently I can work around this by adding some kind of attribute which says "threadIdx.x maps to the corresponding element in the output tensor", and then directly putting `B[threadIdx.x]` in `StorageFlatten`.

**Comment by tqchen** [comment]
We should already handle similar logics， maybe try to change the order you fuse multiple axis cc @sxjscience

**Comment by tqchen** [comment]
My guess is that you can already get what you want by using a different way to fuse all axis(we can do it either from first to last, or from last to first) @phisiart

**Comment by phisiart** [comment]
I tried fusing from the other direction (last to first), but it's still not getting simplified to the most.

This time I got `B[((((threadIdx.x/4)*2)*2) + (threadIdx.x % 4))] = A[((((threadIdx.x/4)*2)*2) + (threadIdx.x % 4))]`.

In `src/lang/buffer.cc`, when simplifying the index expression, it is searching for a pattern like `(e / m) * m + e % m`, and change that to `e`.

When I fuse from first to last, I get `((x / 4) * 2 + (x / 2) % 2) * 2 + x % 2`, but when I fuse from last to first, I get `x / 4 * 2 * 2 + x % 4`. Neither of the expressions has the desired form for simplification.

I think that rather than improving and depending on expression simplification (which we can guarantee to be sound but not necessarily complete, especially when we might have vars and imms mixed together it would be even more complex), we should do something similar to what Halide does: when we know we are in a shader, we don't create a `Store` stmt but a special `Call::image_store` stmt (which will later be transformed to `Call::glsl_texture_store` [see here](https://github.com/halide/Halide/blob/master/src/StorageFlattening.cpp#L247).

**Comment by tqchen** [comment]
@phisiart That is a possible solution that we can adopt

**Comment by phisiart** [comment]
Update: the first set of ops are done. Now `nnvm/tutorials/define_and_compile_model.py`'s resnet runs successfully in local OpenGL.

File: include/tvm/ir.h [hunk 1]
```diff
@@ -226,6 +226,8 @@ constexpr const char* channel_write_advance = "channel_write_advance";
 constexpr const char* pipeline_stage_scope = "pipeline_stage_scope";
 /*! \brief pipeline execution scope, implies the scope can be pipelined. */
 constexpr const char* pipeline_exec_scope = "pipeline_exec_scope";
+/*! \brief This stage is an OpenGL shader. */
+constexpr const char* stage_is_opengl = "stage_is_opengl";
```
Comment (tqchen) [inline include/tvm/ir.h]: just do opengl_stage_scope

Comment (tqchen) [reply include/tvm/ir.h]: add more comment on the behavior implication of such stage, such as every provide is a texture store directly

**Comment by tqchen** [comment]
Can you try to see if we can run the resnet on webGL rpc? The only thing that we have to do likely, is to include graph_runtime.cc in https://github.com/dmlc/tvm/tree/master/src/runtime/graph into web runtime

**Pull request merged**